File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_break_hist_test.ts
/**
 * g_hshk_break_hist_get（IO→CPU コマンド 17h）
 * 根拠: HandShake.mdc「ブレイク履歴取得」
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ handshake で 1 ケースを実行する。
 * @param fn 本体
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * IO が HSHK_REQ_1 を上げるまで待つ。
 * @param mock IO モック
 * @param timeoutMs 上限 ms
 */
async function waitReq1(
  mock: IoBoardHandshakeMock,
  timeoutMs = 2000,
): Promise<void> {
  const t0 = Date.now();
  while (mock.bus.HSHK_REQ_1 !== 1) {
    if (Date.now() - t0 > timeoutMs) {
      throw new Error("timeout waiting HSHK_REQ_1");
    }
    await new Promise((r) => setTimeout(r, 1));
  }
}

/**
 * ハンドラを call で呼び、IO→CPU 交換と並行する。
 * @param mock IO モック
 * @param toCpu IO→CPU フレーム
 * @param fromCpu CPU→IO 待ちバイト数
 * @returns CPU→IO 応答
 */
async function callHandler(
  mock: IoBoardHandshakeMock,
  toCpu: Uint8Array,
  fromCpu: number,
): Promise<Uint8Array> {
  const io = mock.exchangeWithCpu(toCpu, fromCpu);
  await waitReq1(mock);
  await session.call("g_handshake_interrupt_handler", {
    registers: { ...BASE_REGS },
  });
  return io;
}

test("17h スロット 8 はヘッダ 0 のあと NG", async () => {
  await withCase(async (s, mock) => {
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x08, 0x00]),
      9,
    );
    expect(Array.from(reply)).toEqual([0, 0, 0, 0, 0, 0, 0, 0, 0x01]);
    s.expectRegisters({ R0: 0, R4: BASE_REGS.R4 });
  });
});

test("17h 履歴未設定は件数 0 で 02h", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0x42, 0x00, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      9,
    );
    expect(reply[0]).toBe(0);
    expect(reply[1]).toBe(0x02);
    expect(reply[2]).toBe(0x42);
    expect(reply[6]).toBe(0x30);
    expect(reply[7]).toBe(0x00);
    expect(reply[8]).toBe(0x02);
    s.expectRegisters({ R0: 0, R4: BASE_REGS.R4 });
  });
});

test("17h 履歴設定で件数 0 なら OK", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      9,
    );
    expect(reply.length).toBe(9);
    expect(reply[0]).toBe(0);
    expect(reply[1]).toBe(0);
    expect(reply[2]).toBe(0xc2);
    expect(reply[3]).toBe(0x04);
    expect(reply[8]).toBe(0x00);
    s.expectRegisters({ R0: 0x00, R4: BASE_REGS.R4 });
  });
});

/**
 * 履歴エントリ 33 語を物理 3F000h 領域へ置く。
 * @param s セッション
 * @param slot 0–7
 * @param index リング index 0–15
 * @param mark 先頭ワード（識別用）
 */
function plantEntry(
  s: Mn1613AsmSession,
  slot: number,
  index: number,
  mark: number,
): void {
  const base = 0x3f000 + slot * 528 + index * 33;
  s.writeWord(base, mark);
  s.writeWord(base + 1, 0x4567);
  s.writeWord(base + 2, 0x89ab);
  s.writeWord(base + 3, 0xcdef);
  s.writeWord(base + 4, 0xa5a5);
  s.writeWord(base + 5, 0x0000);
  for (let i = 6; i < 33; i += 1) {
    s.writeWord(base + i, 0x1000 + i);
  }
}

test("17h 履歴 1 件を新しい順で返す", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META");
    s.writeWord(meta, 1);
    s.writeWord(meta + 1, 1);
    s.writeWord(meta + 2, 0);
    plantEntry(s, 0, 0, 0x0123);
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      8 + 66 + 1,
    );
    expect(reply[0]).toBe(1);
    expect(reply[1]).toBe(0);
    expect((reply[8]! << 8) | reply[9]!).toBe(0x0123);
    expect((reply[16]! << 8) | reply[17]!).toBe(0xa5a5);
    expect(reply[reply.length - 1]).toBe(0x00);
  });
});

test("17h は 2 件を新しい順（後に書いた方から）で返す", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META");
    s.writeWord(meta, 2);
    s.writeWord(meta + 1, 2);
    s.writeWord(meta + 2, 0);
    plantEntry(s, 0, 0, 0x0001);
    plantEntry(s, 0, 1, 0x0002);
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      8 + 66 * 2 + 1,
    );
    expect(reply[0]).toBe(2);
    expect((reply[8]! << 8) | reply[9]!).toBe(0x0002);
    expect((reply[8 + 66]! << 8) | reply[9 + 66]!).toBe(0x0001);
    expect(reply[reply.length - 1]).toBe(0x00);
  });
});

test("17h スロット 7 は slot×528 先のエントリを返す", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x07, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META") + 7 * 3;
    s.writeWord(meta, 1);
    s.writeWord(meta + 1, 1);
    s.writeWord(meta + 2, 0);
    plantEntry(s, 7, 0, 0x7777);
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x07, 0x00]),
      8 + 66 + 1,
    );
    expect(reply[0]).toBe(1);
    expect((reply[8]! << 8) | reply[9]!).toBe(0x7777);
    expect(reply[reply.length - 1]).toBe(0x00);
  });
});

test("17h Bit0=1 は取得後に当該スロットの履歴をクリアする", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META");
    s.writeWord(meta, 1);
    s.writeWord(meta + 1, 1);
    s.writeWord(meta + 2, 1);
    plantEntry(s, 0, 0, 0x0123);
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x01]),
      8 + 66 + 1,
    );
    expect(reply[0]).toBe(1);
    expect(reply[1]).toBe(0x01);
    expect(reply[reply.length - 1]).toBe(0x00);
    expect(s.readWord(meta)).toBe(0);
    expect(s.readWord(meta + 1)).toBe(0);
    expect(s.readWord(meta + 2)).toBe(0);
  });
});

test("17h オーバフロー済みはステータス Bit0 を立てる", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META");
    s.writeWord(meta, 1);
    s.writeWord(meta + 1, 1);
    s.writeWord(meta + 2, 1);
    plantEntry(s, 0, 0, 0x00aa);
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      8 + 66 + 1,
    );
    expect(reply[0]).toBe(1);
    expect(reply[1]).toBe(0x01);
    expect((reply[8]! << 8) | reply[9]!).toBe(0x00aa);
    expect(reply[reply.length - 1]).toBe(0x00);
  });
});

test("17h 履歴 16 件を新しい順で返す", async () => {
  await withCase(async (s, mock) => {
    await callHandler(
      mock,
      Uint8Array.from([
        0x10, 0x00, 0xc2, 0x04, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00,
      ]),
      1,
    );
    const meta = s.wordAddr("GL_BP_HIST_META");
    s.writeWord(meta, 16);
    s.writeWord(meta + 1, 16);
    s.writeWord(meta + 2, 0);
    for (let i = 0; i < 16; i += 1) {
      plantEntry(s, 0, i, 0x0100 + i);
    }

    const entrySize = 66;
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x17, 0x00, 0x00]),
      8 + entrySize * 16 + 1,
    );

    expect(reply.length).toBe(8 + entrySize * 16 + 1);
    expect(reply[0]).toBe(16);
    expect(reply[1]).toBe(0);
    expect((reply[8]! << 8) | reply[9]!).toBe(0x010f);
    expect((reply[8 + entrySize * 15]! << 8) | reply[9 + entrySize * 15]!).toBe(
      0x0100,
    );
    expect(reply[reply.length - 1]).toBe(0x00);
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_break_hist_test.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const HDR: usize = 10;
const ENTRY_SIZE: usize = 66;

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s.max_cycles = Some(250_000_000);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn wait_req1(mock: &IoBoardHandshakeMock, timeout: Duration) -> Result<(), FrameworkError> {
    let start = Instant::now();
    loop {
        if mock.wires.lock().expect("wires lock").hshk_in_req == 1 {
            return Ok(());
        }
        if start.elapsed() > timeout {
            return Err(FrameworkError::invalid_argument(
                "timeout waiting hshk_in_req",
            ));
        }
        std::thread::yield_now();
    }
}

fn break_set_frame(slot: u8, flags: u8, count: u8, addr: u32, data: u16) -> Vec<u8> {
    vec![
        0x10,
        slot,
        flags,
        count,
        ((addr >> 24) & 0xff) as u8,
        ((addr >> 16) & 0xff) as u8,
        ((addr >> 8) & 0xff) as u8,
        (addr & 0xff) as u8,
        ((data >> 8) & 0xff) as u8,
        (data & 0xff) as u8,
    ]
}

fn call_handler(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    to_cpu: &[u8],
    from_cpu_len: usize,
) -> Result<Vec<u8>, FrameworkError> {
    let io = Arc::clone(mock);
    let frame = to_cpu.to_vec();
    let worker = std::thread::spawn(move || {
        let mut poll = || std::thread::yield_now();
        io.exchange_with_cpu(&frame, from_cpu_len, &mut poll)
    });

    wait_req1(mock, Duration::from_millis(5000))?;
    let _ = s.call(
        "g_handshake_interrupt_handler",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;

    worker
        .join()
        .map_err(|_| FrameworkError::invalid_argument("io worker panicked"))?
}

fn plant_entry(s: &mut Mn1613AsmSession, slot: u8, index: u8, mark: u16) {
    let base = 0x3f000u32 + (slot as u32) * 132 + (index as u32) * 33;
    s.write_word_phys(base, mark);
    s.write_word_phys(base + 1, 0x4567);
    s.write_word_phys(base + 2, 0x89ab);
    s.write_word_phys(base + 3, 0xcdef);
    s.write_word_phys(base + 4, 0xa5a5);
    s.write_word_phys(base + 5, 0x0000);
    for i in 6..33u32 {
        s.write_word_phys(base + i, 0x1000 + i as u16);
    }
}

#[test]
fn break_hist_slot4_returns_header_zero_then_ng() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_handler(s, &mock, &[0x17, 0x04], HDR + 1)?;
        assert_eq!(reply, vec![0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x01]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn break_hist_without_hist_setting_returns_count0_and_status_02() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0x42, 0x00, 0x0000_3000, 0x0000),
            1,
        )?;
        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + 1)?;

        assert_eq!(reply[0], 0);
        assert_eq!(reply[1], 0x42);
        assert_eq!(reply[2], 0x00);
        assert_eq!(reply[3], 0x00);
        assert_eq!(reply[6], 0x30);
        assert_eq!(reply[7], 0x00);
        assert_eq!(reply[8], 0);
        assert_eq!(reply[9], 0);
        assert_eq!(reply[10], 0x02);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn break_hist_with_hist_setting_and_zero_count_returns_ok() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + 1)?;

        assert_eq!(reply.len(), HDR + 1);
        assert_eq!(reply[0], 0);
        assert_eq!(reply[1], 0xc2);
        assert_eq!(reply[2], 0x04);
        assert_eq!(reply[3], 0x00);
        assert_eq!(reply[8], 0);
        assert_eq!(reply[9], 0);
        assert_eq!(reply[10], 0x00);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn break_hist_returns_one_entry_in_order() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let meta = s.word_addr("GL_BP_HIST_META")?;
        s.write_word(meta, 1);
        s.write_word(meta + 1, 1);
        s.write_word(meta + 2, 0);
        plant_entry(s, 0, 0, 0x0123);

        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + ENTRY_SIZE + 1)?;
        assert_eq!(reply[0], 1);
        assert_eq!(reply[1], 0xc2);
        assert_eq!(reply[8], 1);
        assert_eq!(((reply[HDR] as u16) << 8) | reply[HDR + 1] as u16, 0x0123);
        assert_eq!(
            ((reply[HDR + 8] as u16) << 8) | reply[HDR + 9] as u16,
            0xa5a5
        );
        assert_eq!(reply[reply.len() - 1], 0x00);
        Ok(())
    })
}

#[test]
fn break_hist_returns_two_entries_in_order() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let meta = s.word_addr("GL_BP_HIST_META")?;
        s.write_word(meta, 2);
        s.write_word(meta + 1, 2);
        s.write_word(meta + 2, 0);
        plant_entry(s, 0, 0, 0x0001);
        plant_entry(s, 0, 1, 0x0002);

        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + ENTRY_SIZE * 2 + 1)?;
        assert_eq!(reply[0], 2);
        assert_eq!(((reply[HDR] as u16) << 8) | reply[HDR + 1] as u16, 0x0001);
        assert_eq!(
            ((reply[HDR + ENTRY_SIZE] as u16) << 8) | reply[HDR + ENTRY_SIZE + 1] as u16,
            0x0002
        );
        assert_eq!(reply[reply.len() - 1], 0x00);
        Ok(())
    })
}

#[test]
fn break_hist_slot3_reads_entry_at_slot_offset() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(3, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let meta = s.word_addr("GL_BP_HIST_META")?.wrapping_add(3 * 3);
        s.write_word(meta, 1);
        s.write_word(meta + 1, 1);
        s.write_word(meta + 2, 0);
        plant_entry(s, 3, 0, 0x7777);

        let reply = call_handler(s, &mock, &[0x17, 0x03], HDR + ENTRY_SIZE + 1)?;
        assert_eq!(reply[0], 1);
        assert_eq!(((reply[HDR] as u16) << 8) | reply[HDR + 1] as u16, 0x7777);
        assert_eq!(reply[reply.len() - 1], 0x00);
        Ok(())
    })
}

#[test]
fn break_hist_overflow_keeps_meta_and_returns_entries() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let meta = s.word_addr("GL_BP_HIST_META")?;
        s.write_word(meta, 1);
        s.write_word(meta + 1, 1);
        s.write_word(meta + 2, 1);
        plant_entry(s, 0, 0, 0x00aa);

        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + ENTRY_SIZE + 1)?;
        assert_eq!(reply[0], 1);
        assert_eq!(reply[1], 0xc2);
        assert_eq!(s.read_word(meta + 2), 1);
        assert_eq!(((reply[HDR] as u16) << 8) | reply[HDR + 1] as u16, 0x00aa);
        assert_eq!(reply[reply.len() - 1], 0x00);
        Ok(())
    })
}

#[test]
fn break_hist_returns_four_entries_in_order() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let _ = call_handler(
            s,
            &mock,
            &break_set_frame(0, 0xc2, 0x04, 0x0000_3000, 0x0000),
            1,
        )?;
        let meta = s.word_addr("GL_BP_HIST_META")?;
        s.write_word(meta, 4);
        s.write_word(meta + 1, 4);
        s.write_word(meta + 2, 0);
        for i in 0..4u8 {
            plant_entry(s, 0, i, 0x0100 + i as u16);
        }

        let reply = call_handler(s, &mock, &[0x17, 0x00], HDR + ENTRY_SIZE * 4 + 1)?;
        assert_eq!(reply.len(), HDR + ENTRY_SIZE * 4 + 1);
        assert_eq!(reply[0], 4);
        assert_eq!(reply[1], 0xc2);
        assert_eq!(reply[8], 4);
        assert_eq!(((reply[HDR] as u16) << 8) | reply[HDR + 1] as u16, 0x0100);
        assert_eq!(
            ((reply[HDR + ENTRY_SIZE * 3] as u16) << 8) | reply[HDR + ENTRY_SIZE * 3 + 1] as u16,
            0x0103
        );
        assert_eq!(reply[reply.len() - 1], 0x00);
        Ok(())
    })
}
